### PR TITLE
Update GPU backend (bugfixes, LLVM 3.7 support)

### PR DESCRIPTION
**Pull request by gkanwar**
The updated backend uses the approach taken by Halide: an NVPTX backend with LLVM passes to emit a PTX string, instead of calling out to NVVM. This should allow us to keep pace with updating LLVM versions.

There are still two test suites that fail for specific reasons:
-- ffi: We don't have GPU ffi support
-- strings: String intrinsics are not implemented for the GPU yet

We also still do not generate reasonable GPU code for matrix multiplication, since the emitted IR is designed for serial execution.

### CODE REVIEW

**Comment by gkanwar** [comment]
[jenkins] test this please.

**Comment by gkanwar** [comment]
[jenkins] test this please. (trying to update the Jenkins config to build with LLVM 3.7)

**Comment by gkanwar** [comment]
Looks like the LLVM backend passes with 3.7! I think this is ready for review at this point.

We should update the Jenkins master builders to 3.7 as well once this is merged.

**Comment by rennis250** [comment]
Confirmed to build and run on Mac 10.11.5 with Xcode V7.3 with LLVM 3.7.1 from Homebrew. All tests pass and all example programs check. I don't have an NVidia card, so can't check GPU support. The springs example successfully builds and runs, but one needs to set two environment variables first, which I didn't notice documented anywhere.

Not sure if I test multi-cpu mode or only a single processor though. How can I be sure that simit is running in multi-cpu mode? Does it activate this automatically if you have more than one CPU?

Also checked that all of the documentation successfully builds and opens. :)

**Comment by rennis250** [comment]
However, running "fem" with the fem_linear.sim program fails:

```
./fem fem_linear.sim ../data/tet-bunny/bunny.1
Internal error at /Users/rje/simit/src/backend/backend_visitor.cpp:23 in visitError
 IndexedTensor should never reach the backend: i
libc++abi.dylib: terminating with uncaught exception of type simit::SimitException:
[1]    98530 abort      ./fem fem_linear.sim ../data/tet-bunny/bunny.1
```

Best,
Rob

File: src/backend/llvm/llvm_util.cpp [hunk 1]
```diff
@@ -28,4 +34,32 @@ std::ostream &operator<<(std::ostream &os, const llvm::Module &module) {
...
+  uint64_t addr = ee->getFunctionAddress(func->getName());
+  iassert(addr != 0)
+      << "MCJIT prevents modifying the module after ExecutionEngine code "
+      << "generation. Ensure all functions are created before fetching "
+      << "function addresses.";
+  return reinterpret_cast<VoidFuncPtr>(addr);
+}
+
+
+void maybeLogModule(llvm::Module *module, std::string fileName) {
```
Comment (fredrikbk) [inline src/backend/llvm/llvm_util.cpp]: Just call the logModule? Log functions are understood to maybe ignore stuff.

File: src/ir.cpp [hunk 2]
```diff
@@ -757,7 +757,9 @@ Stmt CallStmt::make(std::vector<Var> results,
 
 // struct Scope
 Stmt Scope::make(Stmt scopedStmt) {
-  iassert(scopedStmt.defined());
+  if (!scopedStmt.defined()) {
```
Comment (fredrikbk) [inline src/ir.cpp]: Is this really necessary? Looks like it just papers over mistaken logic in building IR? Do leave it in if it's really convenient though!

File: src/ir_codegen.cpp [hunk 3]
```diff
@@ -12,79 +12,82 @@ using namespace std;
 namespace simit {
 namespace ir {
 
+Expr getZeroVal(const TensorType *type) {
```
Comment (fredrikbk) [inline src/ir_codegen.cpp]: Nice, it's good to see code deleted!

**Comment by fredrikbk** [comment]
This pull request makes llvm 3.7 or higher mandatory. Since Halide does the same I think it's fine, but it will cause some churn for users so we'll have to do some work to announce it (I can take care of this).

**Comment by fredrikbk** [comment]
I got some errors on my system with llvm 3.8.1 so I have set up a branch where I can work on them (https://github.com/simit-lang/simit/tree/llvm-3.7-plus-support).

**Comment by fredrikbk** [comment]
@gkanwar: This pull request has 53 commits, but most of them seems to be re-commits of mine and Stephen's previous commits. Do you know why this happens?

File: src/backend/llvm/llvm_util.cpp (hunk 1 above)
Comment (gkanwar) [inline src/backend/llvm/llvm_util.cpp]: Good point, updated in #47

File: src/ir.cpp (hunk 2 above)
Comment (gkanwar) [inline src/ir.cpp]: Yup, looks like this was a remnant of a previous iteration of a GPU lowering pass I was working on. Reintroduced the assert in #47

**Comment by gkanwar** [comment]
@rennis250 -- Thank you for the feedback! It looks like the fem app issue existed before this change, so we will look into it separately. (Side node: I do believe the README.md file in the springs app discusses the environment variables, but you're right that it's not very prominently mentioned!)

@fredrikbk -- I created a new branch which fixed the rebasing artifacts and fixed some of the issues you pointed out. Let's move discussion over to #47. Thanks!

**Pull request closed without merging**